### PR TITLE
Async search: rename REST parameters

### LINKED ISSUE #54069: Clarify async search REST parameters
When playing with the new async search API I noticed a couple of inconsistencies and potential naming problems that I would like to discuss. Note that it's important to address these now as the API haven't been released yet and they are declared stable in our REST spec.

I asked @karmi for his input to validate my concerns and come up with some proposal. The following are the problems and the changes that we are proposing:

- `wait_for_completion`: it indicates how long you are willing to block and wait for results when submitting an async search, effectively turning async search to sync.
`wait_for_completion` is used in other API but with type `boolean`, while it is exposed as a `number`, effectively a timeout, in submit async search. This introduces inconsistency in our REST API, and it will cause issues for some of the language clients.
**Proposal**: rename it to `wait_for_results_timeout`: this way we include the timeout terminology and we don't reuse the existing `wait_for_completion`. Also `results` better explains what it is that users are waiting for compared to `completion`.

- `keep_alive`: it indicates how long the async search is available within the cluster. That means that when such timeout expires, the search will be stopped if still running or its results will be purged if it has already completed. 
The `keep_alive` naming comes from http terminology where it has to do with connections, while here the semantics is around how long state will be available/stored in the cluster, which could lead to misinterpreting what the parameter does.
**Proposal**: rename it to `keep_results_timeout`: this way we move away from reusing http terminology, and we make it clear that it's also a timeout around how long results will be available. Maybe what is not super clear about this is that the counting starts when the async search is submitted, not when it is completed. Suggestions are welcome.

- `clean_on_completion`: it indicates whether results should not be stored once they are returned within the above described timeout. 
There is some double negation in its description that makes it hard to understand it. Also, the notion of `completion` can be confusing as it's not about whether the search was completed but whether the results were returned within the provided (currently `wait_for_completion`) timeout. By default, results are not stored when they are returned directly by submit async search. Being it a `boolean` it may make users think that they can disable storing results at all times, but storing results can not be disabled, rightly so, when submit async search did not return them within the timeout. 
I considered removing this parameter, because when results have been returned, they could be stored externally. It turns out though that this parameter is useful to make testing deterministic and it makes sense to keep it.
**Proposal**: rename it to `keep_results` and make it an `enum` rather than a `boolean` with two possible values: `auto` (the default behaviour: store results unless submit async search returned them within `keep_results_timeout`) and `always` (store results for later retrieval even if they have been returned by submit async search within the provided timeout). I find that this better reflects the behaviour of the API and aligns well with the above proposed rename of `keep_alive` to `keep_results_timeout` as they are somehow related. Note that `always` does not mean forever, the results will always be cleaned when their validity expires.

- The rename of `wait_for_completion` to `wait_for_results_timeout` should also be applied to the get async search API, but maybe we should consider whether this parameter is useful when retrieving results? Users that are calling get async search are taking advantage of the async nature of async search, hence while I see why one would block and wait when submitting, I don't see why one would block and wait when retrieving results. Is avoiding an additional call when the search is almost complete a good enough reason to expose this parameter?

### PR DESCRIPTION
**Pull request by javanna**
This commit renames `wait_for_completion` to `wait_for_completion_timeout` in submit async search and get async search.
Also it renames `clean_on_completion` to `keep_on_completion` and turns around its behaviour.

Closes #54069

### CODE REVIEW

**Comment by bpintea** [comment]
7.7 has been FFrozen, can the `v7.7.0` label be removed?

**Comment by javanna** [comment]
@Mpdreamz can correct me if I am wrong, but I think that this a blocker for the language clients. As far as I understand `wait_for_completion` is expected to be a `boolean` like in other API and breaks some of the clients, which is why it was decided to rename it.

**Comment by bpintea** [comment]
Thanks, @javanna 

> @Mpdreamz can correct me if I am wrong, but I think that this a blocker for the language clients.

@jpountz May I update the list of blocking issues to include this one?

**Comment by Mpdreamz** [comment]
It won't break the clients (unless it was an enum) but these type of changes are incredibly hard to patch after a release. Since this was discussed and agreed upon by a large group  on the issueI would vote to raise it as blocker so that it can go in the next bc. Similar to #54196

**Comment by jpountz** [comment]
+1 to merge to 7.7. I replaced the `>enhancement` label with `>non-issue` since async search has not been released yet.

File: docs/reference/search/async-search.asciidoc [hunk 1]
```diff
@@ -193,11 +193,12 @@ first.
...
-The `wait_for_completion` parameter, which defaults to `1`, can also be provided
-when calling the Get Async Search API, in order to wait for the search to be
-completed up until the provided timeout. Final results will be returned if
-available before the timeout expires, otherwise the currently available results
-will be returned once the timeout expires.
+The `wait_for_completion_timeout` parameter can also be provided when calling
+the Get Async Search API, in order to wait for the search to be completed up
+until the provided timeout. Final results will be returned if available before
+the timeout expires, otherwise the currently available results will be
+returned once the timeout expires. By default no timeout is set meaning that
```
Comment (jimczi) [inline docs/reference/search/async-search.asciidoc]: The default is set to `1s` currently.

File: x-pack/plugin/src/test/resources/rest-api-spec/api/async_search.get.json [hunk 2]
```diff
@@ -21,15 +21,13 @@
       ]
     },
     "params":{
-      "wait_for_completion":{
+      "wait_for_completion_timeout":{
         "type":"time",
-        "description":"Specify the time that the request should block waiting for the final response",
-        "default": "1s"
+        "description":"Specify the time that the request should block waiting for the final response"
```
Comment (jimczi) [inline x-pack/plugin/src/test/resources/rest-api-spec/api/async_search.get.json]: the default is still 1s ?

**Review by jimczi — COMMENTED** [review]
I left two comments, LGTM otherwise

**Review by jimczi — APPROVED** [review]

File: docs/reference/search/async-search.asciidoc (hunk 1 above)
Comment (javanna) [reply docs/reference/search/async-search.asciidoc]: for get? Am I misreading the default in the request?

Comment (jimczi) [reply docs/reference/search/async-search.asciidoc]: ah right only the submit has the default, sorry.

File: x-pack/plugin/src/test/resources/rest-api-spec/api/async_search.get.json (hunk 2 above)
Comment (jimczi) [reply x-pack/plugin/src/test/resources/rest-api-spec/api/async_search.get.json]: same as above

**Comment by javanna** [comment]
heads up @lukasolson you will have to rename `waitForCompletion` to `waitForCompletionTimeout` in your code. I believe Kibana is not using `cleanOnCompletion` hence no change needed there.

**Comment by lizozom** [comment]
@lukasolson @javanna was this backported to 7.7?

**Comment by javanna** [comment]
@lizozom yes it was, see https://github.com/elastic/elasticsearch/commit/28142261a22265f913d072089fce93921bad993c

**Comment by lizozom** [comment]
@javanna I'm asking, because backporting the kibana change to 7.7 fails in testing, while the 7.x and the 8.0 branches were working and merged 

https://github.com/elastic/kibana/pull/61917

**Comment by lizozom** [comment]
@javanna I think I might have found the issue. Lets see how the build goes!

**Comment by javanna** [comment]
@lizozom I see, let me know if I can do anything to help.

**Pull request merged**